src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -483,72 +483,100 @@ quota_t cap_to_ceiling(quota_t& value, const quota_t limit) {
     return 0;
 }
 
-/// If \p delta is negative, splits its value accourding to \p surplus values:
-/// - amount under total surplus is split pro rata to \p surplus amounts
-/// - amount over total surplus is split equally. Values in \p schedule
-/// are added to.
-/// \pre schedule.size() == surplus.size()
+/// If \p delta is negative, splits its value pro rata to \p quotas values and
+/// into the slices limited by \p quota values. The slices are added to the
+/// elements of \p schedule.
+/// \pre sum(\p quotas) < abs(delta)
+/// \pre schedule.size() == quotas.size()
+/// \pre 0 <= quotas[i] for any i
 void dispense_negative_deltas(
   std::vector<quota_t>& schedule,
-  quota_t& delta,
-  const std::vector<quota_t>& surplus) {
+  quota_t delta,
+  std::vector<quota_t>&& quotas) {
+    vassert(
+      schedule.size() == quotas.size(),
+      "Schedule and quotas sizes must be equal");
```

Comment:
is this a case where we are certain this to be true at the risk of hard stopping a broker in production?

Reply:
this is the case, unless the number of cores can change on the fly. I've done that to be able to address the vectors by `[]` instead of `at()`. But for the peace of mind, I agree let's have an error log instead.

Reply:
yeh the number of cores will not change at runtime.